cirq/qis/states.py

```diff
@@ -36,6 +37,503 @@
+        """
+        is_state_vector = self.data.shape == (self._dim,)
+        is_state_tensor = self.data.shape == self.qid_shape
+        if is_state_vector or is_state_tensor:
+            validate_normalized_state_vector(
+                self.state_vector(), qid_shape=self.qid_shape, dtype=dtype, atol=atol
+            )
+        elif self._is_density_matrix():
+            validate_density_matrix(
+                self.density_matrix(), qid_shape=self.qid_shape, dtype=dtype, atol=atol
+            )
+        else:
+            raise ValueError(
+                'Invalid quantum state: '
+                f'Data shape of {self.data.shape} is not '
+                f'compatible with qid shape of {self.qid_shape}.'
+            )
+
+
+def quantum_state(
+    state: 'cirq.QUANTUM_STATE_LIKE',
+    qid_shape: Optional[Tuple[int, ...]] = None,
+    *,  # Force keyword arguments
+    copy: bool = False,
+    validate: bool = True,
+    dtype: Optional[Type[np.number]] = None,
+    atol: float = 1e-7,
+) -> QuantumState:
+    """Create a QuantumState object from a state-like object.
+
+    Args:
+        state: The state-like object.
+        qid_shape: The qid shape.
+        copy: Whether to copy the data underlying the state.
+        validate: Whether to check if the given data and qid shape
+            represent a valid quantum state with the given dtype.
+        dtype: The desired data type.
+        atol: Absolute numerical tolerance to use for validation.
+
+    Raises:
+        ValueError: Invalid quantum state.
+        ValueError: The qid shape was not specified and could not be inferred.
+    """
+    if isinstance(state, QuantumState):
+        if qid_shape is not None and state.qid_shape != qid_shape:
+            raise ValueError(
+                'The specified qid shape must be the same as the '
+                'qid shape of the given state.\n'
+                f'Specified shape: {qid_shape}\n'
+                f'Shape of state: {state.qid_shape}.'
+            )
+        if copy or dtype and dtype != state.dtype:
+            if dtype and dtype != state.dtype:
+                data = state.data.astype(dtype, casting='unsafe', copy=True)
+            else:
+                data = state.data.copy()
+            new_state = QuantumState(data, state.qid_shape)
+        else:
+            new_state = state
+        if validate:
+            new_state.validate(dtype=dtype, atol=atol)
+        return new_state
+
+    if isinstance(state, value.ProductState):
+        actual_qid_shape = (2,) * len(state)
+        if qid_shape is not None and qid_shape != actual_qid_shape:
+            raise ValueError(
+                'The specified qid shape must be the same as the '
+                'qid shape of the given state.\n'
+                f'Specified shape: {qid_shape}\n'
+                f'Shape of state: {actual_qid_shape}.'
+            )
+        if dtype is None:
+            dtype = np.complex64
```

Comment:
can we pull this out to a constant, like DEFAULT_COMPLEX_PRECISION? It feels like a decision that requires shotgun surgery to extract. I know it exists elsewhere in the codebase, but maybe we could eventually get all of them to a single point?

Reply:
Done.